### ISSUE TITLE
Microsoft.Extensions.DependencyInjection and Service Remoting with Stateless Services

### ISSUE THREAD

**Issue by Mardoxx** [Azure/service-fabric-issues#269]
How can I use Microsoft's DI framework (or any other DI framework) with Service Fabric's remoting library?

Let's say I have an `IService` which relies on an EF `DbContext` implementation. How can I set up DI so that for each "request" to the service a new `DbContext` is created. I naively injected it in to the `IService` constructor which, of course, doesn't work since this is only created once, and not called per request.

What would be an appropriate setup for this? What would I have to do to achieve this?

Basically I'd like it so that when my `IService`'s methods are called (resolved) new instances of registered services are created appropriately.

**Comment by Mardoxx** [comment]
For information: https://alexmg.com/introducing-the-autofac-integration-for-service-fabric/

**Comment by vturecek** [comment]
@Mardoxx You'd have to introduce a second level of indirection, where each IService method is simply a wrapper around a class of your own that you can instantiate and inject dependencies into. 

The class the implements IService is your service class, and its lifetime represents the lifetime of a service instance/replica. It is not like an MVC controller where you get a new instance to process each request.

**Comment by masnider** [comment]
Hi @Mardoxx - please let us know if Vas's suggestions helped and if you were able to resolve your issue. If we don't hear back from you in a few days with more information, we'll probably close this issue down for now.

**Comment by Mardoxx** [comment]
Apologise, I thought I'd replied and closed this. He definitely pointed me in the right direction!!

I managed to solve this using Autofac - which can set this all up for you.

I did it by injecting `Func<Owned<IService>> serviceFactory`. Then per method call you can do something like

    using (var scope = _serviceFactory())
    {
        var service = scope.Value;
        await service.DoSomethingAsync();
    }

I don't mind taking the dependency on AutoFac, in fact I quite like it (!), but it would be nice to see some sort of recommended-Microsoft-way of doing this in some sample somewhere in the future :)

**Comment by masnider** [comment]
Great!

**Comment by NArnott** [comment]
Would be be possible to move to a more MVC controller approach so that a class/method could use dependency injection similarly? 

I've not been a fan of the class that implements IService having to be the service class. It would be great to separate that out and allow IOC at the same time.
